Added configuration for AppService

davidreneuw opened the pull request:
Configurations can now be selected in the toolbox page for any tool that requires it, and HandleTerraformRequestServiceAsync can now handle an optional configuration object, which is added to the Project_Resource2 object's JsonContent attribute associated with the request.

There are preset configurations that can be used (for supported frameworks), or custom configurations (for personal docker compose)

The TerraformOutputHandler also now triggers the configuration pipeline on ADO when an AppService is finished deploying using the configuration in the JsonContent

There will be associated PRs for the datahub-resource-modules repo and we will have to decide how we want to host the different template frameworks

FYI: this is not the final stage of this branch, I want to add more testing and implement feedback

davidreneuw committed:
Added AppServiceConfiguration, and passing it around up to terraform and to the post terraform triggers

yjmrobert commented on Portal/src/Datahub.Core/Services/Docs/DocumentationService.cs:
```diff
@@ -52,7 +52,7 @@ public class DocumentationService
         var branch = environment.IsProduction()? "main": "next";
         _docsRoot = config.GetValue(DOCS_ROOT_CONFIG_KEY, $"https://raw.githubusercontent.com/ssc-sp/datahub-docs/{branch}/")!;
-        _docsEditPrefix = config.GetValue(DOCS_EDIT_URL_CONFIG_KEY, "https://github.com/ssc-sp/datahub-docs/edit/{branch}/")!;
+        _docsEditPrefix = config.GetValue(DOCS_EDIT_URL_CONFIG_KEY, $"https://github.com/ssc-sp/datahub-docs/edit/{branch}/")!;
```
Nice catch!

yjmrobert commented on Portal/src/Datahub.Core/Services/Projects/IRequestManagementService.cs:
```diff
@@ -6,7 +6,7 @@
 using Datahub.Core.Model.Projects;
 using Datahub.Shared.Entities;
-
+#nullable enable
```
Let's enable nullable on the whole Core csproj

yjmrobert commented on Portal/src/Datahub.Functions/AzureConfig.cs:
```diff
@@ -84,4 +84,6 @@ public class AdoConfig
     public string ProjectName { get; set; } = "FSDH SSC";
     public string URL { get; set; } = "https://dev.azure.com/{organization}/{project}/_apis/wit/workitems/${workItemTypeName}?api-version=6.0";
+    public string PipelineURL { get; set; } = "https://dev.azure.com/{organization}/{project}/_apis/pipelines/{pipelineId}/runs?api-version=7.1-preview.1";
+    public string AppServiceConfigPipelineId { get; set; } = "9";
```
Is there a way to retrieve this off a name? Maybe a query? Not sure but it would be nice to have it more dynamic

yjmrobert commented on Portal/src/Datahub.Functions/TerraformOutputHandler.cs:
Is there a way we can omit the keyvault client and simply save the secret names?

I'd like to move that piece with the POST request into a separate function so it doesn't trip up any other TF outputs

yjmrobert commented on Portal/src/Datahub.Core/Services/Projects/IRequestManagementService.cs:
```diff
@@ -21,7 +21,7 @@ public interface IRequestManagementService
     /// <returns>A task that represents the asynchronous operation. The task result contains a boolean indicating whether the Terraform request was handled successfully or not.</returns>
     Task<bool> HandleTerraformRequestServiceAsync(Datahub_Project project, string terraformTemplate,
-        PortalUser requestingUser);
+        PortalUser requestingUser, object? configuration = null);
```
Can the configuration be pulled out of the workspace definition instead of passing it as a parameter? I'm not convinced this is the right approach

yjmrobert commented on Portal/src/Datahub.ProjectTools/Services/RequestManagementService.cs:
```diff
@@ -74,13 +78,21 @@ await ctx.Entry(project)
+        {
+            _logger.LogInformation("Required configuration not provided for {TerraformTemplate}", requestedTemplate.Name);
+            return;
+        }
```
Why allow a defaulted nullable parameter only to return? I think I'm missing something

yjmrobert commented on Portal/src/Datahub.ProjectTools/Services/RequestManagementService.cs:
```diff
@@ -74,13 +78,21 @@ await ctx.Entry(project)
             ResourceType = TerraformTemplate.GetTerraformServiceType(requestedTemplate.Name)
         };
 
+        resource.JsonContent = ProcessConfiguration(resource, configuration);
```
This should be handled by the functions or portal to represent the current state of the resource. Let's move the user inputs into the `InputJsonContent` column

yjmrobert commented on Portal/src/Datahub.ProjectTools/Services/RequestManagementService.cs:
```diff
@@ -145,6 +158,31 @@ await ctx.Entry(project)
+        }
+
+        return jsonContent;
+    }
```
See my previous comment, I think we can move this into the dialog box in the workspace toolbox

yjmrobert commented on Portal/src/Datahub.Portal/Views/Dialogs/AzureAppServiceConfigurationDialog.razor:
I'd love to see a screenshot of this component

davidreneuw replied:
I dont think you'd love it.... it is super basic and I haven't really played with it yet, it was written in a crunch time lol

yjmrobert replied:
So true lol

yjmrobert commented on Portal/src/Datahub.Portal/Pages/Workspace/Toolbox/WorkspaceToolboxPage.razor:
```diff
@@ -220,6 +229,35 @@
+        }
+
+        return null;
+    }
```
What do you think about moving the app configuration to an App Service Page instead of the Toolbox?

yjmrobert commented on Shared/src/Datahub.Shared/Entities/AppServiceConfiguration.cs:
Great idea putting it here, I'm not entirely sure it needs to be a shared file at the moment - but there are discussions around making an API endpoint for the resource provisioner to also handle configuration so TBD

yjmrobert commented on Shared/src/Datahub.Shared/Entities/WorkspaceAppData.cs:
```diff
@@ -1,6 +1,7 @@
 namespace Datahub.Shared.Entities;
-
+#nullable enable
 public class WorkspaceAppData
 {
         public string DatabricksHostUrl { get; set; } = string.Empty;
+        public AppServiceConfiguration? AppServiceConfiguration { get; set; }
```
Just set it to a default instead of nullable for the shared project

davidreneuw committed:
Created App Service Page where configuration now happens. Stopped passing configuration around and extracted it from workspace definition instead. Implemented other various changes.

davidreneuw committed:
Missing in last commit

davidreneuw commented:
Web App page, not configured:
![NotConfigured](https://github.com/ssc-sp/datahub-portal/assets/56747050/d80d7f74-ae75-4c70-bf43-57503175a069)

When configured:
![Configured](https://github.com/ssc-sp/datahub-portal/assets/56747050/a2d47859-452c-42e8-99e8-fffd253bdf8d)

Configuration dialog:
![Configuration](https://github.com/ssc-sp/datahub-portal/assets/56747050/303c0c5a-5b69-4815-854b-83582a08095a)

What I hope to add in the next few commits:
 - Testing configuring app services
 - Providing live app service status (ping url?)
 - configuration validation and testing

yjmrobert commented on Shared/src/Datahub.Shared/Entities/AppServiceConfiguration.cs:
```diff
@@ -0,0 +1,46 @@
+{
+    public class AppServiceConfiguration
+    {
+        public string AppServiceFramework { get; set; } = string.Empty;
+        public string AppServiceGitRepo { get; set; } = string.Empty;
+        public string AppServiceComposePath { get; set; } = string.Empty;
+        public string AppServiceId { get; set; } = string.Empty;
+        public string AppServiceHostName { get; set; } = string.Empty;
+        public string AppServiceRg { get; set; } = string.Empty;
```
Just a thought that you can probably omit the `AppService` bit from the nested variables 😄 

Since it'll be read as something like `workspaceAppServiceConfiguration.GitRepo`

yjmrobert replied:
and then you can expand `Rg` into `ResourceGroupName` 😉

yjmrobert commented on Portal/src/Datahub.ProjectTools/Services/RequestManagementService.cs:
Don't forgot to clean up the summaries!

yjmrobert commented on Portal/src/Datahub.Portal/Views/Dialogs/AzureAppServiceConfigurationDialog.razor:
```diff
@@ -0,0 +1,42 @@
+﻿@using Datahub.Shared.Entities
+
+<MudDialog>
+    <DialogContent>
+        <MudSelect Label="Select a framework" @bind-Value="selectedFramework">
+            @foreach(string framework in AppServiceTemplates.TEMPLATES)
+            {
+                <MudSelectItem Value="@framework" Label="@framework" />
+            }
+        </MudSelect>
+        <MudTextField Label="Enter git repo URL" @bind-Value="gitRepoUrl" Disabled="@(selectedFramework!=AppServiceTemplates.CUSTOM)"/>
+
+        <MudTextField Label="Enter Docker compose file path" @bind-Value="dockerComposeFilePath" Disabled="@(selectedFramework!=AppServiceTemplates.CUSTOM)"/>
+    </DialogContent>
+    <DialogActions>
+        <MudButton Color="Color.Primary" Variant="Variant.Filled" OnClick="Submit">Submit</MudButton>
+    </DialogActions>
+</MudDialog>
```
Localize! Don't be like me and forgot about i18n haha

yjmrobert commented on Portal/src/Datahub.Portal/Pages/Workspace/WebApp/WorkspaceWebAppPage.razor:
```diff
@@ -0,0 +1,184 @@
+        }
+        catch (MicrosoftIdentityWebChallengeUserException e)
+        {
+            _consentHandler.HandleException(e);
```
Which scope are you adding here?

davidreneuw replied:
Its just an artifact of "copilot, create a web app page based and inspire yourself from the database page"

yjmrobert commented on Portal/src/Datahub.Infrastructure/Queues/MessageHandlers/AppServiceConfigurationSender.cs:
```diff
@@ -0,0 +1,14 @@
+        {
+        }
+
+        protected override string ConfigPathOrQueueName => "app-service-configuration";
```
Can you add a prefix like `workspace-app-service-configuration`, just to avoid any possible confusion with the portal or other

yjmrobert commented on Portal/src/Datahub.Infrastructure/Queues/Messages/AppServiceConfigurationMessage.cs:
```diff
@@ -0,0 +1,8 @@
+
+namespace Datahub.Infrastructure.Queues.Messages
+{
+    public record AppServiceConfigurationMessage(
+        string ProjectAcronym
+        ) : IRequest;
```
Is the message to update the workspace configuration off the database values or are we setting them via messages?

Either way it might make it clearer to add a verb in the object name around that, ie `WorkspaceAppServiceConfigurationRefreshMessage` or `WorkspaceAppServiceConfigurationUpdateMessage`

It's a bit verbose but it might clarify the usage and intent a bit

yjmrobert commented on Portal/src/Datahub.Functions/ConfigureAppService.cs:
```diff
@@ -0,0 +1,156 @@
+        
+        private async Task PostPipelineRun(string pipelineUrl, AppServiceConfiguration appServiceConfiguration, string projectAcronym, string rgName)
+        {
+            var body = new
+            {
+                resources = new
+                {
+                    repositories = new
+                    {
+                        self = new
+                        {
+                            refName = "refs/heads/main"
+                        }
+                    }
+                },
+                templateParameters = new
+                {
+                    resourceGroup = rgName,
+                    webAppId = appServiceConfiguration.AppServiceId,
+                    gitUrl = appServiceConfiguration.AppServiceGitRepo,
+                    composePath = appServiceConfiguration.AppServiceComposePath
+                }
+            };
```
You're gonna love this snippet I keep saved as a refresher

## Easy way to work with JSON and C#
```csharp
var postdata = new JsonObject
{
    ["moduleList"] = new JsonObject
    {
        ["modules"] = new JsonArray(new JsonObject()
        {
            ["moduleArea"] = "Discovery",
            ["moduleType"] = "ChannelListing",
            ["moduleRequest"] = new JsonObject
            {
                ["consumeRequests"] = new JsonArray(),
                ["resultTemplate"] = "responsive",
                ["alerts"] = new JsonArray(),
                ["profileInfos"] = new JsonArray()
            }
        })
    }
};
```

yjmrobert commented on Portal/src/Datahub.Functions/AzureConfig.cs:
```diff
@@ -84,4 +84,6 @@ public class AdoConfig
     public string OrgName { get; set; } = "DataSolutionsDonnees";
     public string ProjectName { get; set; } = "FSDH SSC";
     public string URL { get; set; } = "https://dev.azure.com/{organization}/{project}/_apis/wit/workitems/${workItemTypeName}?api-version=6.0";
+    public string GetPipelineURL { get; set; } = "https://dev.azure.com/{organization}/{project}/_apis/pipelines?api-version=7.1-preview.1";
```
Can update these variable names to `WorkItemsUrlTemplate` and `ListPipelinesUrlTemplate`

yjmrobert commented on Portal/src/Datahub.Core/Utils/TerraformVariableExtraction.cs:
```diff
@@ -48,13 +48,72 @@ public static class TerraformVariableExtraction
+    /// <returns>The AppServiceConfiguration object containing the configuration info of the app service</returns>
+    public static AppServiceConfiguration ExtractAppServiceConfiguration(Datahub_Project? project)
+    {
+        var appServiceTemplateName = TerraformTemplate.GetTerraformServiceType(TerraformTemplate.AzureAppService);
+        var appServiceFramework = ExtractStringVariable(
+            project?.Resources?.FirstOrDefault(r => r.ResourceType == appServiceTemplateName)?.InputJsonContent,
+            "app_service_framework");
+        var appServiceGitRepo = ExtractStringVariable(
+            project?.Resources?.FirstOrDefault(r => r.ResourceType == appServiceTemplateName)?.InputJsonContent,
+            "app_service_git_repo");
+        var appServiceComposePath = ExtractStringVariable(
+            project?.Resources?.FirstOrDefault(r => r.ResourceType == appServiceTemplateName)?.InputJsonContent,
+            "app_service_compose_path");
+        var appServiceId = ExtractStringVariable(
+            project?.Resources?.FirstOrDefault(r => r.ResourceType == appServiceTemplateName)?.JsonContent,
+            "app_service_id");
+        var appServiceHostName = ExtractStringVariable(
+            project?.Resources?.FirstOrDefault(r => r.ResourceType == appServiceTemplateName)?.JsonContent,
+            "app_service_host_name");
+        var appServiceRg = ExtractStringVariable(
+            project?.Resources?.FirstOrDefault(r => r.ResourceType == appServiceTemplateName)?.JsonContent,
+            "app_service_rg");
+        return new AppServiceConfiguration(appServiceFramework, appServiceGitRepo,
+            appServiceComposePath, appServiceId, appServiceHostName, appServiceRg);
+    }
```
I would just find the project resource and return the method call below

yjmrobert reviewed (commented):
Really good work, coming along really nicely!

I can't wait to see this released, going to blow our users away for sure.

A minor comment on the gear button in the screenshots, I would turn it into a Primary Button with a label (configure? settings? both?) as well as the gear icon - just to drive attention there

davidreneuw committed:
Refactored some naming, added a web app dashboard card and other

davidreneuw commented:
Made requested changes and added a dashboard web app card:
![image](https://github.com/ssc-sp/datahub-portal/assets/56747050/c5e0b9fc-94cf-416a-aa38-27701507da18)

davidreneuw committed:
Merge branch 'develop' into dr/web-app-configuration

davidreneuw committed:
Fixing merge conflict

yjmrobert commented on Portal/src/Datahub.Functions/AzureConfig.cs:
```diff
@@ -83,5 +83,7 @@ public class AdoConfig
     public string OrgName { get; set; } = "DataSolutionsDonnees";
     public string ProjectName { get; set; } = "FSDH SSC";
-    public string URL { get; set; } = "https://dev.azure.com/{organization}/{project}/_apis/wit/workitems/${workItemTypeName}?api-version=6.0";
+    public string WorkItemsUrlTemplate { get; set; } = "https://dev.azure.com/{organization}/{project}/_apis/wit/workitems/${workItemTypeName}?api-version=6.0";
```
Is the `$` in `${workItemTypeName} part of the url? I was going to edit and remove it directly but then I thought it might be some requirement in the ADO url?

davidreneuw replied:
[Its required in the url](https://learn.microsoft.com/en-us/rest/api/azure/devops/wit/work-items/create?view=azure-devops-rest-7.1&tabs=HTTP)

yjmrobert commented on Portal/src/Datahub.Portal/Pages/Workspace/Dashboard/AppServiceCard.razor:
```diff
@@ -0,0 +1,48 @@
+        </MudElement>
+        <MudElement HtmlTag="div">
+            <MudButton
+                Href=@($"/w/{WorkspaceAcronym}/webapp")
```
`PageRoutes.WorkspacePrefix` and lets make one for the web app path as well

yjmrobert commented on Portal/src/Datahub.Portal/Pages/Workspace/Dashboard/AppServiceCard.razor:
```diff
@@ -0,0 +1,48 @@
+                Target="_blank"
+                Color="Color.Primary"
+                Variant="@Variant.Text"
+                EndIcon="@Icons.Material.Outlined.OpenInNew">
```
I've been moving over to the font awesome fa-light version, check out the file `SidebarIcons` for some static references.

Bit of a misnomer as originally it was for the sidebar, but having all the icon references in one spot for consistency is nice

yjmrobert commented on Portal/src/Datahub.Portal/Pages/Workspace/Dashboard/AppServiceCard.razor:
```diff
@@ -0,0 +1,48 @@
+
+    private string GetWebAppUrl()
+    {
+        return TerraformVariableExtraction.ExtractAppServiceConfiguration(ProjectResource).HostName;
```
Possible null reference exception here

yjmrobert commented on Portal/src/Datahub.Portal/Pages/Workspace/WebApp/WorkspaceWebAppPage.razor:
```diff
@@ -0,0 +1,186 @@
+        @if (_isProvisioned)
+        {
+            <MudButton Color="Color.Primary" Size="Size.Medium" Variant="Variant.Filled" OnClick="async () => { await ConfigureWebApp(); }">
+                @Localizer["Configure "]
+                <MudIcon Icon="@Icons.Material.Filled.Settings"/>
```
Instead of the ` ` on the localization I think you can achieve the same effect by putting the `ml-2` class on the icon

yjmrobert commented on Portal/src/Datahub.Portal/Pages/Workspace/WebApp/WorkspaceWebAppPage.razor:
```diff
@@ -0,0 +1,186 @@
+        @Localizer["Here you can manage your web application. You can view and edit the configurations of your web application"]
+    </MudText>
+
+    <MudText Typo="Typo.h2">
+        @Localizer["Web Application Info"]
+    </MudText>
+    @if (_isProvisioned)
+    {
+        <MudText>
+            @Localizer["Host: "] <MudLink Href="@_webAppHost">@_webAppHost</MudLink>
+        </MudText>
+        <MudText>
+            @Localizer["Id: "] <code>@_webAppId</code>
+        </MudText>
+        <MudText>
+            @Localizer["Status: "]
+        </MudText>
+        @if (_isConfigured)
+        {
+            <MudText>
+                @Localizer["Framework: "] <code>@_webAppFramework</code>
+            </MudText>
+            <MudText>
+                @Localizer["Git repo: "] <code>@_webAppRepo</code>
+            </MudText>
+            <MudText>
+                @Localizer["Compose path: "] <code>@_webAppCompose</code>
+            </MudText>
+        }
```
Nice, check out the updated simple table format on the database page, makes it all nice and clean

yjmrobert commented on Portal/src/Datahub.Portal/Pages/Workspace/WebApp/WorkspaceWebAppPage.razor:
```diff
@@ -0,0 +1,186 @@
+
+@code {
+
+    [Parameter] public string WorkspaceAcronym { get; set; }
```
`[Parameter, EditorRequired]` will throw a build exception if someone forgets it somehow

yjmrobert commented on Portal/src/Datahub.Portal/Pages/Workspace/WebApp/WorkspaceWebAppPage.razor:
```diff
@@ -0,0 +1,186 @@
+
+        var appConfiguration = TerraformVariableExtraction.ExtractAppServiceConfiguration(workspace);
+
+        _webAppHost = appConfiguration.HostName;
+        _webAppId = appConfiguration.Id;
```
Possible null reference on `appConfiguration`?

yjmrobert commented on Portal/src/Datahub.ProjectTools/Services/RequestManagementService.cs:
```diff
@@ -10,6 +12,7 @@
 using Microsoft.Extensions.Logging;
 using Newtonsoft.Json;
 using System.Transactions;
+using AngleSharp.Common;
```
?

yjmrobert replied:
lol

davidreneuw committed:
🔧 Beef up configuration dialog and web app page. Add tests and validation. Switch icon away from material icon. Fix some namespace differences in functions test

davidreneuw committed:
Merge branch 'develop' into dr/web-app-configuration

# Conflicts:
#	Portal/src/Datahub.Functions/AzureConfig.cs
#	Portal/src/Datahub.Functions/BugReport.cs
#	Portal/src/Datahub.Portal/Layout/SidebarIcons.cs

davidreneuw committed:
✨ Added environment variable tables. These can be editable and they interface with the workspace's keyvault.

✨ Added web app management (start, stop and restart) from web app page

✨ other quality of life features!

davidreneuw committed:
Merge branch 'develop' into dr/web-app-configuration

davidreneuw committed:
🔧 Show key values as code

davidreneuw committed:
Merge remote-tracking branch 'origin/dr/web-app-configuration' into dr/web-app-configuration

davidreneuw commented:
**Pictures:**
Dashboard card:
![image](https://github.com/ssc-sp/datahub-portal/assets/56747050/24d88abc-992b-417e-ada3-554f24db41cb)

Workspace web app page, viewed as an admin:
![image](https://github.com/ssc-sp/datahub-portal/assets/56747050/71f670b5-4d95-4772-8cbe-5a65eb26aa27)

Stopping the web app:
![image](https://github.com/ssc-sp/datahub-portal/assets/56747050/c7cce83f-f132-4309-87bd-f4604fb6e227)

The environment variables table has two mode, editing and viewing. In viewing mode, a toggle appears to show/hide env vars values:
![image](https://github.com/ssc-sp/datahub-portal/assets/56747050/80b21a28-f94d-46ce-bb29-2bfe67f5acbd)

All icons that have no text with them have a label attached to them.

Configuration dialog. Entering a PAT in Git URL field will automatically turn git repo visibility to private and move the PAT over:
![image](https://github.com/ssc-sp/datahub-portal/assets/56747050/ba510eb9-5bd9-46cd-8cea-45187b6ae558)

Selecting a template instead of a custom docker compose:
![image](https://github.com/ssc-sp/datahub-portal/assets/56747050/3deac79a-1df2-453f-a68b-7673425b187f)

Configuration dialog contains an editable environment variables table:
![image](https://github.com/ssc-sp/datahub-portal/assets/56747050/f2ca7abd-1c1d-4ea9-b97d-d301da421ba4)

**Architecture**:
The workspace web app interacts with the workspace web app management service. This service, given the web app id, is responsible for starting, stopping, resetting and configuring the web app, in addition to interacting with the app service project resource to save user input from configuration dialog.

Web app configuration is triggered by the web app management service, which builds a configuration message that is sent to the ConfigureWorkspaceAppService function. This function creates an ADO client to find the right pipeline to use, then makes a request to run that pipeline with all the proper information.

The ADO pipeline clones the given repo, builds the docker-compose and pushes it, then sends it to the app service as the container settings.

This process could likely be simplified.

**Known issues or missing features:**
- When editing the environment variables table, as soon as one of the fields is clicked on, the commit edit and cancel edit buttons disappear
- Nothing is done currently with the PAT. I imagine the goal is to essentially save it in KV then pass it to the function on configuration, and rebuild the git url to include the PAT when sending to ADO pipeline
- Need to show proxy details

yjmrobert commented on Portal/src/Datahub.Functions/ConfigureWorkspaceAppService.cs:
```diff
@@ -0,0 +1,177 @@
+        private readonly ILogger<ConfigureWorkspaceAppService> _logger;
+        private readonly AzureConfig _config;
+        private readonly IDbContextFactory<DatahubProjectDBContext> _dbContext;
+        internal HttpClient _httpClient = new();
```
It's good practice to create `HttpClient` as needed, similar to the database context

yjmrobert commented on Portal/src/Datahub.Portal/Components/EnvironmentVariablesTable.razor.cs:
```diff
@@ -0,0 +1,159 @@
+        private void HandleCommitEditClicked(MouseEventArgs args)
+        {
+            _logger.LogInformation("Commit edit button clicked.");
+            _snackbar.Add(Localizer["Sending environment variable updated"], Severity.Info);
```
"Sending environment variable update"

yjmrobert commented on Portal/src/Datahub.Portal/Datahub.Portal.csproj:
```diff
@@ -10,8 +10,22 @@
+	</ItemGroup>
+	<ItemGroup>
+		<None Include="..\.editorconfig" Link=".editorconfig" />
+	</ItemGroup>
```
Don't need these I don't think

yjmrobert commented on Portal/src/Datahub.Portal/Datahub.Portal.csproj:
```diff
@@ -10,8 +10,22 @@
+		<PackageReference Include="Azure.ResourceManager.AppService" Version="1.0.2" />
 		<PackageReference Include="Azure.ResourceManager.PostgreSql" Version="1.1.1" />
 		<PackageReference Include="DocumentFormat.OpenXml" Version="2.20.0" />
+		<PackageReference Include="Microsoft.Data.SqlClient" Version="5.1.4" />
```
What's this for?

davidreneuw replied:
Merge conflict artifact

yjmrobert commented on Portal/src/Datahub.Functions/TerraformOutputHandler.cs:
```diff
@@ -178,16 +175,20 @@ internal async Task ProcessAzureWebApp(IReadOnlyDictionary<string, TerraformOutp
         projectResource.CreatedAt = DateTime.UtcNow;
         projectResource.JsonContent = jsonContent.ToString();
 
+        await _projectDbContext.SaveChangesAsync();
```
no need to save changes twice

davidreneuw replied:
Thanks for catching that!

yjmrobert commented on Portal/src/Datahub.Portal/Components/EnvironmentVariablesTable.razor:
```diff
@@ -0,0 +1,168 @@
+            <MudToggleIconButton @bind-Toggled="@_show" Icon="@SidebarIcons.Show" ToggledIcon="@SidebarIcons.Hide" Title="@Localizer["Show environment variable values"]" ToggledTitle="@Localizer["Hide environment variable values"]"/>
+        </ToolBarContent>
+
+        <ColGroup>
+            <col/>
+            <col/>
+            <col/>
+        </ColGroup>
```
I think we can remove this? I believe it defaults to something like this if nothing is presented? Or does this have a visual impact on the table

davidreneuw replied:
Was originally going to make a column for a "show value"/"hide value" button for each row to show/hide single rows instead of the full table but scrapped that halfway through. Will remove this, it is not needed.

yjmrobert commented on Portal/src/Datahub.Portal/Pages/Workspace/Dashboard/AppServiceCard.razor:
```diff
@@ -0,0 +1,50 @@
+    [Parameter] [EditorRequired] public string WorkspaceAcronym { get; set; }
+
+
+    private string GetWebAppUrl()
+    {
+        var webAppConfig = TerraformVariableExtraction.ExtractAppServiceConfiguration(ProjectResource);
+        var url = new UriBuilder(webAppConfig.HostName);
+        return url.Uri.ToString();
+    }
```
I think this needs to use the proxy, @ErikApption can you confirm the semantic pathing for workspace apps?

yjmrobert replied:
just saw in the other PR, it's `/webapp/<acronym>/(**catch-all)`

yjmrobert commented on Portal/src/Datahub.Portal/Pages/Workspace/WebApp/WorkspaceWebAppInfoTable.razor:
```diff
@@ -0,0 +1,75 @@
+    
+    protected override void OnParametersSet()
+    {
+        _url = new UriBuilder(WebAppHost).Uri.ToString();
```
update to semantic proxy path

yjmrobert commented on Portal/src/Datahub.Portal/Startup.cs:
```diff
@@ -405,6 +407,9 @@ private void ConfigureCoreDatahubServices(IServiceCollection services)
             services.AddScoped<UpdateProjectMonthlyCostService>();
             services.AddScoped<IProjectCreationService, ProjectCreationService>();
+
+            services.AddSingleton<IWorkspaceWebAppManagementService, WorkspaceWebAppManagementService>();
```
Change to `AddScoped`, we don't want any sharing across sessions

yjmrobert commented on Portal/src/Datahub.Portal/Views/Dialogs/AzureAppServiceConfigurationDialog.razor:
```diff
@@ -0,0 +1,132 @@
+
+                <MudTextField Label=@Localizer["Enter Docker compose file path"]   @bind-Value="_dockerComposeFilePath" HelperText="@Localizer["Path from the base of your repo to the desired compose file, including the compose file"]" Placeholder="path/to/compose.yaml" Required="true" RequiredError=@Localizer["Path to docker compose is necessary, file name must be included"]/>
+            }
+            <MudExpansionPanels Elevation="1" Class="mt-4">
+                <MudExpansionPanel IsInitiallyExpanded="false">
+                    <TitleContent>
+                        <MudStack Row="true">
+                            <MudIcon Icon=@SidebarIcons.Brackets/>
+                            <MudText>
+                                @Localizer["Environment variables"]
+                            </MudText>
+                        </MudStack>
+                    </TitleContent>
+                    <ChildContent>
+                        <MudText>
+                            @Localizer["Environment variables are key-value pairs that are used to pass configuration to your web application. They are used to store sensitive information such as database connection strings, API keys, etc. You can add environment variables to your web application by providing a key-value pair. The key is the name of the environment variable and the value is the v ...
+                        </MudText>
+                        <EnvironmentVariablesTable projectAcronym="@WorkspaceAcronym" resource="@Resource" isEditable="true" @ref="@envVarTable"/>
+                    </ChildContent>
+                </MudExpansionPanel>
+            </MudExpansionPanels>
```
What is the thinking behind collapsing this? Are there a lot of them in our users projects?

davidreneuw replied:
Did not want to scare users with lots of stuff as soon as the dialog opens. This way users who don't use env vars can just ignore it.

(I also think it looks better that way)

yjmrobert replied:
Gotcha, sounds good

yjmrobert commented on Portal/test/Datahub.Functions.UnitTests/EmailServiceTests.cs:
```diff
@@ -1,8 +1,8 @@
+﻿using Datahub.Functions.Services;
 using Microsoft.Extensions.Logging;
-using Datahub.Infrastructure.Queues.Messages;
+using Moq;
```
?

davidreneuw replied:
Can't say I know where this comes from or why to be honest, will revert

yjmrobert commented:
Looks good! Can you bring the web-app-configuration pipeline into the `infra` folder?

yjmrobert commented:
Just to follow-up

> When editing the environment variables table, as soon as one of the fields is clicked on, the commit edit and cancel edit buttons disappear

Is this in the MudDialog? I wonder if it's an interaction within that.

> Nothing is done currently with the PAT. I imagine the goal is to essentially save it in KV then pass it to the function on configuration, and rebuild the git url to include the PAT when sending to ADO pipeline

Yeah, at first glance something like a `WorkspaceAppPatSecretName` with `"workspace-app-pat"` in the AKV comes to mind. The issue around that is we're now exposing that user's PAT to the others in the workspace. Might have to think a bit more around safeguards and edge cases for that one

davidreneuw commented:
> Is this in the MudDialog? I wonder if it's an interaction within that.

Yea it is, I'll test that out in the main page to see if thats solely due to the dialog

> Yeah, at first glance something like a `WorkspaceAppPatSecretName` with `"workspace-app-pat"` in the AKV comes to mind. The issue around that is we're now exposing that user's PAT to the others in the workspace. Might have to think a bit more around safeguards and edge cases for that one

I'll do that for now. We could have it that any role can store secrets but only admins can read secrets, to limit the amount of access to values.

davidreneuw commented:
@yjmrobert Just FYI even in main page the behaviour still happens:
![image](https://github.com/ssc-sp/datahub-portal/assets/56747050/5577741e-5cd8-4023-a31d-a944d2b8653c)
![image](https://github.com/ssc-sp/datahub-portal/assets/56747050/92ea34cf-453f-489f-9604-f863057c1525)

Weird because the component is essentially the same as your IP whitelist table... I'll have to keep looking

davidreneuw committed:
🔧 Implement minor fixes provided

yjmrobert commented:
Looks good to me, lmk when you find the bug

ErikApption committed:
merged develop
